compiler/src/test/java/com/squareup/anvil/compiler/InterfaceMergerTest.kt

```diff
@@ -622,4 +623,221 @@ class InterfaceMergerTest(
+      $import
+      
+      @ContributesTo(Any::class)
+      @ContributesTo(Unit::class)
+      interface ContributingInterface
+      
+      @ContributesTo(Any::class, replaces = [ContributingInterface::class])
+      interface SecondContributingInterface
+      
+      $annotation(Any::class)
+      interface ComponentInterface
+      
+      $annotation(Unit::class)
+      interface SubcomponentInterface
+      """
+    ) {
+      assertThat(componentInterface extends contributingInterface).isFalse()
+      assertThat(componentInterface extends secondContributingInterface).isTrue()
+      assertThat(subcomponentInterface extends contributingInterface).isTrue()
+      assertThat(subcomponentInterface extends secondContributingInterface).isFalse()
+    }
+  }
+
+  @Test
+  fun `replaced interfaces contributed to multiple scopes must use the same scope`() {
+    assumeIrBackend()
+
+    compile(
+      """
+      package com.squareup.test
+      
+      import com.squareup.anvil.annotations.ContributesTo
+      $import
+      
+      @ContributesTo(Any::class)
+      @ContributesTo(Unit::class)
+      interface ContributingInterface
+      
+      @ContributesTo(Int::class, replaces = [ContributingInterface::class])
+      interface SecondContributingInterface
+      
+      $annotation(Any::class)
+      interface ComponentInterface
+      
+      $annotation(Unit::class)
+      interface SubcomponentInterface1
+      
+      $annotation(Int::class)
+      interface SubcomponentInterface2
+      """
+    ) {
+      assertThat(exitCode).isError()
+      assertThat(messages).contains(
+        "com.squareup.test.SecondContributingInterface with scope kotlin.Int wants to " +
+          "replace com.squareup.test.ContributingInterface, but the replaced class isn't " +
+          "contributed to the same scope."
+      )
+    }
+  }
+
+  @Test fun `interfaces contributed to multiple scopes can be excluded in one scope`() {
+    assumeIrBackend()
+
+    compile(
+      """
+      package com.squareup.test
+      
+      import com.squareup.anvil.annotations.ContributesTo
+      $import
+      
+      @ContributesTo(Any::class)
+      @ContributesTo(Unit::class)
+      interface ContributingInterface
+      
+      @ContributesTo(Any::class)
+      interface SecondContributingInterface
+      
+      $annotation(Any::class, exclude = [ContributingInterface::class])
+      interface ComponentInterface
+      
+      $annotation(Unit::class)
+      interface SubcomponentInterface
+      """
+    ) {
+      assertThat(componentInterface extends contributingInterface).isFalse()
+      assertThat(componentInterface extends secondContributingInterface).isTrue()
+      assertThat(subcomponentInterface extends contributingInterface).isTrue()
+      assertThat(subcomponentInterface extends secondContributingInterface).isFalse()
+    }
+  }
+
+  @Test fun `contributed interfaces in the old format are picked up`() {
```

Comment:
Couple questions on this one:
Could you clarify what the old format is here? (is it referring to the second file in the first compilation?
Why do we disable Anvil for the first compilation?

Reply:
In the old format we generated two properties `reference` and `scope` for each contribution. In the new format we generate `reference` and `scope1, scope2, scope3,...`. This test makes sure that we still find a scope without an index number. 

That's why we must disable Anvil in the first compilation. If it would be enabled, then we would generate hints in the new format. That's not what we want to test. That's why I disabled Anvil and wrote the code manually that Anvil would have generated before.

Reply:
Ahh that makes sense, thanks for explaining 👍